### PR TITLE
Initial work to add support for external language service result providers for…

### PR DESCRIPTION
**Pull request by mavasani**
…Find All References, GotoDefinition and NavigateTo.

All the newly added result provider interfaces are internal, and would eventually go through API design review to enable third party plugin implementations.

### CODE REVIEW

**Comment by mavasani** [comment]
test vsi please

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-analysis @srivatsn @heejaechang @mattwar @DustinCampbell @CyrusNajmabadi @Pilchie for review.

We are hoping to get this change ported to future-stabilization before EOD tomorrow.

**Comment by heejaechang** [comment]
adding @davkean since I am not sure whether he is okay adding more code churn in future-stabilization.

File: src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs [hunk 1]
```diff
@@ -59,6 +59,16 @@ private static IEnumerable<Location> GetPreferredSourceLocations(ISymbol symbol)
...
         }
 
+        public static IEnumerable<INavigableItem> GetPreferredNavigableItems(Solution solution, IEnumerable<INavigableItem> navigableItems)
+        {
+            var generatedCodeRecognitionService = solution.Workspace.Services.GetService<IGeneratedCodeRecognitionService>();
+            navigableItems = navigableItems.Where(n => n.Document != null);
+            var hasNonGeneratedCodeItem = navigableItems.Any(n => !generatedCodeRecognitionService.IsGeneratedCode(n.Document));
+            return hasNonGeneratedCodeItem ?
+                navigableItems.Where(n => !generatedCodeRecognitionService.IsGeneratedCode(n.Document)) :
+                navigableItems.Where(n => generatedCodeRecognitionService.IsGeneratedCode(n.Document));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs]: I don't like this.  Currently we filter out results from generated code, and it makes the experience in Roslyn itself very unpleasant.  Specifically, trying to navigate to syntax node types is super hard to do.  I don't seewhy we want to filter things out.

File: src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs [hunk 2]
```diff
@@ -59,6 +59,16 @@ private static IEnumerable<Location> GetPreferredSourceLocations(ISymbol symbol)
                 : locations.Where(loc => loc.IsInSource);
         }
 
+        public static IEnumerable<INavigableItem> GetPreferredNavigableItems(Solution solution, IEnumerable<INavigableItem> navigableItems)
+        {
+            var generatedCodeRecognitionService = solution.Workspace.Services.GetService<IGeneratedCodeRecognitionService>();
+            navigableItems = navigableItems.Where(n => n.Document != null);
+            var hasNonGeneratedCodeItem = navigableItems.Any(n => !generatedCodeRecognitionService.IsGeneratedCode(n.Document));
+            return hasNonGeneratedCodeItem ?
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs]: preferred IDE style is:

```
return x
    ? trueCase
    : falseCase;
```

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs [hunk 3]
```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
             return null;
         }
 
-        public async Task<IEnumerable<INavigableItem>> FindReferencesAsync(Document document, int position, IWaitContext waitContext)
+        /// <summary>
+        /// Finds references using the externally defined <see cref="IFindReferencesResultProvider"/>s.
+        /// </summary>
+        private async Task AddExternalReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, CancellationToken cancellationToken)
+        {
+            if (_externalReferencesProviders.Any())
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: this line seems to have no effect.

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs [hunk 4]
```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
...
+            {
+                foreach (var provider in _externalReferencesProviders)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: Can we do these operations in parallel instead?

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs [hunk 5]
```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
...
+        private async Task AddSymbolReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, IWaitContext waitContext)
         {
-            var cancellationToken = waitContext.CancellationToken;
             var result = await this.FindReferencedSymbolsAsync(document, position, waitContext).ConfigureAwait(false);
-            if (result == null)
+            if (result != null)
             {
-                return SpecializedCollections.EmptyEnumerable<INavigableItem>();
+                var referencedSymbols = result.Item1;
+                var searchSolution = result.Item2;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: why is this solution necessayr when we have the provided document?

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs [hunk 6]
```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
...
+
+                // TODO: Merging references from SymbolFinder and external providers might lead to duplicate or counter-intuitive results.
+                // TODO: For now, we avoid merging and just display the results either from SymbolFinder or the external result providers but not both.
+                if (builder.Count > 0 && DisplayReferences(builder))
+                {
+                    builder.Free();
+                    return true;
+                }
+
+                builder.Free();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: please use some sort of 'using' construct so we don't have to remember to free on all codepaths.

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs [hunk 7]
```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
...
+            }
+
+            return false;
+        }
+
+        private bool DisplayReferences(IEnumerable<INavigableItem> result)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: bool method should be TryDisplayReferences.

File: src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs (hunk 1 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Extensibility/Navigation/NavigableItemFactory.cs]: I have retained the same generated code based filtering that was already being done for GotoDefinition - I think we at least want to be consistent in both code paths. Can you please open a separate issue for your recommended change?

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs (hunk 4 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: Given that these are MEF based providers, I am not sure if running in parallel can cause any deadlocks and unforeseen side effects. I followed the sequential approach used by CodeFixService to compute fixes - we should definitely think about this aspect when discussing this at the design meeting.

File: src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs (hunk 5 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs]: I have actually just refactored the previous code and moved it within this method - I see that [FindReferencedSymbolsAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/FindReferences/AbstractFindReferencesService.cs,35) might end up mapping the symbol search to a different solution (probably for MAS case).

File: src/EditorFeatures/Core/Implementation/NavigateTo/AbstractNavigateToSearchService.cs [hunk 8]
```diff
@@ -1,134 +1,58 @@
...
-using Microsoft.CodeAnalysis.FindSymbols;
-using Microsoft.CodeAnalysis.Shared.Utilities;
-using Microsoft.VisualStudio.Language.NavigateTo.Interfaces;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo
 {
     internal abstract partial class AbstractNavigateToSearchService : INavigateToSearchService
     {
-        public async Task<IEnumerable<INavigateToSearchResult>> SearchProjectAsync(Project project, string searchPattern, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/NavigateTo/AbstractNavigateToSearchService.cs]: NOTE: All the code here is just moved to a separate file `NavigateToSearchResultProvider`

File: src/EditorFeatures/Core/Implementation/FindReferences/IFindReferencesNavigableItemResultProvider.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Navigation;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    internal interface IFindReferencesResultProvider
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/FindReferences/IFindReferencesNavigableItemResultProvider.cs]: How is this different from the IFindReferencesService service?

File: src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs [hunk 10]
```diff
@@ -77,20 +91,30 @@ internal static class GoToDefinitionHelpers
...
+        public static bool TryGoToDefinition(
+            ImmutableArray<INavigableItem> navigableItems,
+            string title,
+            Workspace workspace,
+            IEnumerable<Lazy<INavigableItemsPresenter>> presenters,
+            bool throwOnHiddenDefinition)
+        {
             // If we have a single location, then just navigate to it.
-            if (preferredSourceLocations.Length == 1)
+            if (navigableItems.Length == 1 && navigableItems[0].Document != null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs]: When we do have navigable items with null documents?  The point of navigable items were that they were effectively a Document+Span pair.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs [hunk 11]
```diff
@@ -123,6 +145,55 @@ internal static class GoToDefinitionHelpers
...
+
+        public static async Task<IEnumerable<INavigableItem>> FindExternalDefinitionsAsync(ISymbol symbol, Project project, IEnumerable<Lazy<INavigableDefinitionProvider>> externalDefinitionProviders, CancellationToken cancellationToken)
+        {
+            foreach (var definitionProvider in externalDefinitionProviders)
+            {
+                var definitions = await definitionProvider.Value.FindDefinitionsAsync(symbol, project, cancellationToken).ConfigureAwait(false);
+                if (definitions != null && definitions.Any())
+                {
+                    var preferredDefinitions = NavigableItemFactory.GetPreferredNavigableItems(project.Solution, definitions);
+                    if (preferredDefinitions.Any())
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs]: this check seems superfluous.  It doesn't change anything.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs [hunk 12]
```diff
@@ -123,6 +145,55 @@ internal static class GoToDefinitionHelpers
...
 
+        public static async Task<IEnumerable<INavigableItem>> FindExternalDefinitionsAsync(Document document, int position, IEnumerable<Lazy<INavigableDefinitionProvider>> externalDefinitionProviders, CancellationToken cancellationToken)
+        {
+            foreach (var definitionProvider in externalDefinitionProviders)
+            {
+                var definitions = await definitionProvider.Value.FindDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+                if (definitions != null && definitions.Any())
+                {
+                    var preferredDefinitions = NavigableItemFactory.GetPreferredNavigableItems(document.Project.Solution, definitions);
+                    if (preferredDefinitions.Any())
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs]: This doesn't seem to have any effect.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/INavigableDefinitionProvider.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        /// <summary>
+        /// Find definitions for the symbol referenced or defined at the given position in the given document.
+        /// </summary>
+        Task<IEnumerable<INavigableItem>> FindDefinitionsAsync(Document document, int position, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Find definitions for the given symbol in the context of the given project.
+        /// </summary>
+        Task<IEnumerable<INavigableItem>> FindDefinitionsAsync(ISymbol symbol, Project project, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/INavigableDefinitionProvider.cs]: I'd prefer these scope themselves in the other direct.  "Project" is larger, so it would come first.

File: src/EditorFeatures/Core/Implementation/GoToImplementation/AbstractGoToImplementationService.cs [hunk 14]
```diff
@@ -17,16 +17,17 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.GoToImplementation
     internal abstract class AbstractGoToImplementationService : IGoToImplementationService
     {
         private readonly IEnumerable<Lazy<INavigableItemsPresenter>> _navigableItemPresenters;
+        private readonly IEnumerable<Lazy<INavigableDefinitionProvider>> _externalDefinitionProviders;
 
-        public AbstractGoToImplementationService(IEnumerable<Lazy<INavigableItemsPresenter>> navigableItemPresenters)
+        public AbstractGoToImplementationService(IEnumerable<Lazy<INavigableItemsPresenter>> navigableItemPresenters, IEnumerable<Lazy<INavigableDefinitionProvider>> externalDefinitionProviders)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/GoToImplementation/AbstractGoToImplementationService.cs]: wrap.

File: src/EditorFeatures/Core/Implementation/NavigateTo/INavigateToSearchResultProvider.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo
+{
+    internal interface INavigateToSearchResultProvider
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/INavigateToSearchResultProvider.cs]: How is this different from the INavigateToSearchService ?

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs [hunk 16]
```diff
@@ -58,6 +58,11 @@ public ReadOnlyCollection<DescriptionItem> DescriptionItems
             private ReadOnlyCollection<DescriptionItem> CreateDescriptionItems()
             {
                 var document = _searchResult.NavigableItem.Document;
+                if (document == null)
+                {
+                    return new List<DescriptionItem>().AsReadOnly();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs]: Can't you just say "new ReadOnlyCollection"?

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs [hunk 17]
```diff
@@ -58,6 +58,11 @@ public ReadOnlyCollection<DescriptionItem> DescriptionItems
             private ReadOnlyCollection<DescriptionItem> CreateDescriptionItems()
             {
                 var document = _searchResult.NavigableItem.Document;
+                if (document == null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs]: When does this happen?

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Navigation;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.VisualStudio.Language.NavigateTo.Interfaces;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo
+{
+    [Export(typeof(INavigateToSearchResultProvider)), Shared]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs]: How does this end up working?   What keeps this from being pulled in for TS projects?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs [hunk 19]
```diff
@@ -237,7 +237,12 @@ private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
             var workspaceRegistration = (WorkspaceRegistration)sender;
             uint docCookie;
-            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+
+            if (!_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie))
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs]: When does this happen?

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSymbolNavigationService.cs [hunk 20]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSymbolNavigationService.cs]: Remove

**Comment by CyrusNajmabadi** [comment]
I'm a bit confused.  The ask is for: to add support for external language service result providers

But don't we have that?  We did that work for TypeScript.  Of course, the APIs are internal.  But looking at your change, everything has stayed internal as well... so i don't see how external people can use this.

File: src/EditorFeatures/Core/Implementation/FindReferences/IFindReferencesNavigableItemResultProvider.cs (hunk 9 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/FindReferences/IFindReferencesNavigableItemResultProvider.cs]: I would say FindReferencesService is analogous to CodeFixService and FindReferencesResultProvider is analogous to CodeFixProvider. The service basically drives all the providers and then is responsible for aggregating/displaying the results. The provider itself should be only data-oriented and eventually we will change these result provider interfaces to abstract types just like CodeFixProvider

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs (hunk 17 above)
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs]: isnt result from outside where this service shouldn't trust?

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs]: Do we not trust the other data as well?  For example, do we verify the source span is within the bounds of hte document we're navigating to?

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs (hunk 18 above)
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs]: not probably in this changeset, but in design meeting, probably want to discuss whether these interface will use pure MEF or something like workspaceservice to tie service to specific workspace + more metadata?

**Comment by heejaechang** [comment]
I believe this change is to add support to chain providers for same language.

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs (hunk 17 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToItemProvider.NavigateToItemDisplay.cs]: Yes, when the API becomes public we would need to check validity for everything within the result returned by external provider. Currently, we have an internal prototype built on top of this result provider API that lazily computes the Document, while requiring to synchronously return the navigable items - we may invariably hit error conditions while lazily computing the document and end up returning a null document. We do ensure that we return valid span though.

In short - the current checks suffice the internal requirements for these APIs but we **should** eventually harden against any value returned by a third party provider when we make the APIs public.

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs (hunk 18 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToSearchResultProvider.cs]: TS has its own navigate to search service http://index/#Microsoft.CodeAnalysis.TypeScript.EditorFeatures/Features/NavigateTo/TypeScriptNavigateToSearchService.cs,12, which is not based on providers. We probably need the approach that HeeJae recommends as the provider probably needs to be able to register for specific workspace(s) only.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs (hunk 12 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs]: We want to return the navigable definitions returned by first definition provider that has at least one preferred item, i.e. non-null Document.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs (hunk 11 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/GoToDefinitionHelpers.cs]: Same here, we want to return results from the first provider returning at least one navigable item.

File: src/EditorFeatures/Core/Implementation/GoToDefinition/INavigableDefinitionProvider.cs (hunk 13 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/GoToDefinition/INavigableDefinitionProvider.cs]: Sure, though we would soon need to clean these up as per design meeting. These should probably look like our CodeFixProvider API: abstract type with RegisterXXX method(s) which take a context parameter that allows registering results.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs (hunk 19 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs]: I hit this while implementing an external workspace that listens to events from IVSRunningDocumentTable4 (just like MiscellaneousFilesWorkspace does) and registers the text buffer for opened document prior to Miscellaneous workspace getting notified of the docCookie - the prior contract assumes that MiscellaneousFilesWorkspace is always the first one to get notified.

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs]: Please comment the code appropriately.

**Comment by mavasani** [comment]
@CyrusNajmabadi I have pushed a new commit addressing your feedback

**Comment by mavasani** [comment]
Ping, can someone please take another look as we want to check this in by tomorrow's future-stabilization cutoff?

To clarify again, this change is just meant to kick off work for supporting 3rd party LS result providers - design of public APIs and code hardening will happen later.

**Comment by mavasani** [comment]
This has been checked into future stabilization and back ported to future - closing this PR.
